### PR TITLE
Product misc endpoints for b2b

### PR DESCRIPTION
**Pull request by OKrivtsova**
product-labels
product-management-attributes
product-measurement-units
product-reviews
product-tax-sets

### CODE REVIEW

File: tests/api/b2b/glue/general_product_endpoints/product_management_attributes/positive.robot [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    And Response body parameter should not be EMPTY:    [data][0][attributes][localizedKeys]
+    And Response body parameter should not be EMPTY:    [data][0][attributes][values]
+    And Response body parameter should not be EMPTY:    [data][0][links][self]
+    Each array element of array in response should contain nested property:    [data]    [attributes]    key
+    Each array element of array in response should contain nested property:    [data]    [attributes]    inputType
+    Each array element of array in response should contain nested property:    [data]    [attributes]    allowInput
+    Each array element of array in response should contain nested property:    [data]    [attributes]    isSuper
+    Each array element of array in response should contain nested property:    [data]    [attributes]    localizedKeys
+    Each array element of array in response should contain nested property:    [data]    [attributes]    values
+    Each array element of array in response should contain nested property:    [data]    [links]    self
```
Comment (AntonShopin) [inline tests/api/b2b/glue/general_product_endpoints/product_management_attributes/positive.robot:28]: Could you please also check that: 
* each 'localizedKeys' array has 'localeName' and 'translation'
* 'localeName' and 'translation' from 'localizedKeys' are not empty 
* each 'values' array has 'value' and 'localizedValues'

General question, is it possible to create a new keyword that will check that value of the parameter is in something?
 Example: Response body parameter should in: [data][attributes][allowInput] true false

File: tests/api/b2b/glue/general_product_endpoints/product_management_attributes/positive.robot [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    Then Response status code should be:    200
+    And Response reason should be:    OK
+    And Response header parameter should be:    Content-Type    ${default_header_content_type}
+    And Response body parameter should be:    [data][id]    ${product_management_attribute_no_input_id}
+    And Response body parameter should be:    [data][type]    product-management-attributes
+    And Response body parameter should be:    [data][attributes][key]    ${product_management_attribute_no_input_id}
+    And Response body parameter should be:    [data][attributes][inputType]    text
+    And Response body parameter should be:    [data][attributes][allowInput]    False
+    And Response body parameter should be:    [data][attributes][isSuper]    False
+    And Response body has correct self link internal
```
Comment (AntonShopin) [inline tests/api/b2b/glue/general_product_endpoints/product_management_attributes/positive.robot:68]: Could you please add one more test, that will check attribute with NOT empty 'localizedKeys'  and 'values'. I'm talking about this part

`"localizedKeys": [
                    {
                        "localeName": "en_US",
                        "translation": "White balance"
                    },
                    {
                        "localeName": "de_DE",
                        "translation": "Weißabgleich"
                    }
                ],
                "values": [
                    {
                        "value": "manual",
                        "localizedValues": [
                            {
                                "localeName": "en_US",
                                "translation": "Manual"
                            },
                            {
                                "localeName": "de_DE",
                                "translation": "Manuell"
                            }
                        ]
                    },`

Example of such attribute: https://glue.de.b2b.demo-spryker.com/product-management-attributes/farbe

Comment (AntonShopin) [inline tests/api/b2b/glue/general_product_endpoints/product_management_attributes/positive.robot:68]: Could you please add one more test, that will check attribute with NOT empty 'localizedKeys'  and 'values'. I'm talking about this part

`"localizedKeys": [
                    {
                        "localeName": "en_US",
                        "translation": "White balance"
                    },
                    {
                        "localeName": "de_DE",
                        "translation": "Weißabgleich"
                    }
                ],
                "values": [
                    {
                        "value": "manual",
                        "localizedValues": [
                            {
                                "localeName": "en_US",
                                "translation": "Manual"
                            },
                            {
                                "localeName": "de_DE",
                                "translation": "Manuell"
                            }
                        ]
                    },`

Example of such attribute: https://glue.de.b2b.demo-spryker.com/product-management-attributes/farbe

File: tests/api/b2b/glue/general_product_endpoints/product_reviews/positive.robot [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    And Response header parameter should be:    Content-Type    ${default_header_content_type}
+    And Response should contain the array of a certain size:    [data]    ${abstract_product_with_reviews_qty}
+    And Response body parameter should be:    [data][id]    ${review_id}
+    And Response body parameter should be:    [data][type]    product-reviews
+    And Response body parameter should not be EMPTY:    [data][attributes][rating]
+    And Response body parameter should not be EMPTY:    [data][attributes][nickname]
+    And Response body parameter should not be EMPTY:    [data][attributes][summary]
+    And Response body parameter should not be EMPTY:    [data][attributes][description]
+    And Response body parameter should not be EMPTY:    [data][links][self]
+
```
Comment (AntonShopin) [inline tests/api/b2b/glue/general_product_endpoints/product_reviews/positive.robot:51]: what about POST requests for reviews?

Comment (AntonShopin) [inline tests/api/b2b/glue/general_product_endpoints/product_reviews/positive.robot:51]: what about POST request for reviews?

File: tests/api/b2b/glue/general_product_endpoints/product_reviews/negative.robot [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    And Response should return error message:    Abstract product sku is not specified.
+
+# bug CC-16486
+Get_a_reviews_with_non_existent_abstract_product
+    When I send a GET request:    /abstract-products/fake/product-reviews/78
+    Then Response status code should be:    404
+    And Response reason should be:    Not Found
+    And Response should return error code:    3402
+    And Response should return error message:    Product review not found.
+
```
Comment (AntonShopin) [inline tests/api/b2b/glue/general_product_endpoints/product_reviews/negative.robot:42]: what about POST requests for reviews?

Comment (AntonShopin) [inline tests/api/b2b/glue/general_product_endpoints/product_reviews/negative.robot:42]: what about POST requests for reviews?

File: tests/api/b2b/glue/general_product_endpoints/product_reviews/positive.robot [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+Get_product_reviews
+    When I send a GET request:    /abstract-products/${abstract_product_with_reviews}/product-reviews
+    Then Response status code should be:    200
+    And Response reason should be:    OK
+    And Response header parameter should be:    Content-Type    ${default_header_content_type}
+    And Response should contain the array of a certain size:    [data]    ${abstract_product_with_reviews_qty}
+    And Response body parameter should not be EMPTY:    [data][0][id]
+    And Response body parameter should be:    [data][0][type]    product-reviews
+    And Response body parameter should not be EMPTY:    [data][0][attributes][rating]
```
Comment (AntonShopin) [inline tests/api/b2b/glue/general_product_endpoints/product_reviews/positive.robot:15]: could you please check that rating is int or decimal (not string :) )

Comment (AntonShopin) [inline tests/api/b2b/glue/general_product_endpoints/product_reviews/positive.robot:15]: could you please check that rating is int or decimal (not string :) )

File: tests/api/b2b/glue/general_product_endpoints/product_reviews/positive.robot (hunk 3 above)
Comment (AntonShopin) [inline tests/api/b2b/glue/general_product_endpoints/product_reviews/positive.robot:51]: Could you please also add test chat will check that it's possible to manage returned number of reviews. 
Example: https://glue.de.b2b.demo-spryker.com/abstract-products/M719/product-reviews?page[offset]=0&page[limit]=1 (only one review returned)

Comment (AntonShopin) [inline tests/api/b2b/glue/general_product_endpoints/product_reviews/positive.robot:51]: Could you please also add test chat will check that it's possible to manage returned number of reviews. 
Example: https://glue.de.b2b.demo-spryker.com/abstract-products/M719/product-reviews?page[offset]=0&page[limit]=1 (only one review returned)

File: resources/environments/environments_api_b2b.json [hunk 6]
```diff
@@ -174,7 +175,16 @@
...
         "concrete_product_multiple_image_set_name": "Wood stackable chair, upholstered - VE 4 pieces, substructure chromed - upholstery black",
         "concrete_available_product_with_stock": "420566",
-        "concrete_available_product_without_stock": ""
+        "concrete_available_product_without_stock": "",
+
+        "concrete_product_with_promotions": "212427",
+
+        "concrete_product_random_weight_sku": "cable-vga-1-2",
+        "concrete_product_random_weight_measurement_unit": "19",
+        "packaging_unit_m": "METR",
```
Comment (AntonShopin) [inline resources/environments/environments_api_b2b.json:184]: are you sure that it's a packaging and not measurement unit?

Comment (AntonShopin) [inline resources/environments/environments_api_b2b.json:184]: are you sure that it's a packaging and not measurement unit?

File: tests/api/b2b/glue/weight_unit_endpoints/product_measurement_units/positive.robot [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+
+Get_product_measurement_unit_by_id
+    When I send a GET request:    /product-measurement-units/${packaging_unit_m}
+    Then Response status code should be:    200
+    And Response reason should be:    OK
+    And Response header parameter should be:    Content-Type    ${default_header_content_type}
+    And Response body parameter should be:    [data][id]    ${packaging_unit_m}
+    And Response body parameter should be:    [data][type]    product-measurement-units
+    And Response body parameter should not be EMPTY:    [data][attributes][name]
+    And Response body parameter should not be EMPTY:    [data][attributes][defaultPrecision]
```
Comment (AntonShopin) [inline tests/api/b2b/glue/weight_unit_endpoints/product_measurement_units/positive.robot:15]: please also check that it's an integer or decimal (not string :) )

Comment (AntonShopin) [inline tests/api/b2b/glue/weight_unit_endpoints/product_measurement_units/positive.robot:15]: please also check that it's an integer

File: tests/api/b2b/glue/general_product_endpoints/product_management_attributes/positive.robot (hunk 1 above)
Comment (OKrivtsova) [reply tests/api/b2b/glue/general_product_endpoints/product_management_attributes/positive.robot:28]: added keywords:
Response body parameter should be in:
Each array element of array in response should contain property with value in:
Each array element of array in response should contain property with value NOT in:
Each array element of nested array should contain property with value in:
Each array element of nested array should contain property with value NOT in:

added checks

File: tests/api/b2b/glue/general_product_endpoints/product_management_attributes/positive.robot (hunk 2 above)
Comment (OKrivtsova) [reply tests/api/b2b/glue/general_product_endpoints/product_management_attributes/positive.robot:68]: added for attributes which have them

Comment (OKrivtsova) [reply tests/api/b2b/glue/general_product_endpoints/product_management_attributes/positive.robot:68]: added for attributes which have them

File: tests/api/b2b/glue/general_product_endpoints/product_reviews/negative.robot (hunk 4 above)
Comment (OKrivtsova) [reply tests/api/b2b/glue/general_product_endpoints/product_reviews/negative.robot:42]: added

File: tests/api/b2b/glue/general_product_endpoints/product_reviews/positive.robot (hunk 3 above)
Comment (OKrivtsova) [reply tests/api/b2b/glue/general_product_endpoints/product_reviews/positive.robot:73]: done

File: tests/api/b2b/glue/general_product_endpoints/product_reviews/positive.robot (hunk 5 above)
Comment (OKrivtsova) [reply tests/api/b2b/glue/general_product_endpoints/product_reviews/positive.robot]: it is always int. added check for int

File: resources/common/common_api.robot [hunk 8]
```diff
@@ -381,6 +381,23 @@ Response body parameter should be:
     Log    ${data}
     Should Be Equal    ${data}    ${expected_value}
 
+Response body parameter should be in:
```
[the remark this answers is not in the document]
Comment (OKrivtsova) [reply resources/common/common_api.robot:384]: done

File: resources/environments/environments_api_b2b.json (hunk 6 above)
Comment (OKrivtsova) [reply resources/environments/environments_api_b2b.json:186]: done

**Review by AntonShopin — APPROVED** [review]
